pkg/services/sqlstore/migrations/correlations_mig.go

```diff
@@ -0,0 +1,21 @@
+package migrations
+
+import (
+	. "github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+)
+
+func addCorrelationsMigrations(mg *Migrator) {
+	correlationsV1 := Table{
+		Name: "correlation",
+		Columns: []*Column{
+			{Name: "uid", Type: DB_NVarchar, Length: 40, Nullable: false, IsPrimaryKey: true},
+			{Name: "source_uid", Type: DB_NVarchar, Length: 40, Nullable: false, IsPrimaryKey: true},
+			// Nullable because in the future we want to have correlations to external resources
+			{Name: "target_uid", Type: DB_NVarchar, Length: 40, Nullable: true},
```

Comment:
There can only be a single primary key, so what you're doing here is adding the primary key {uid, source_uid} which makes lookups that doesn't specify both (or at least not uid) inefficient.

Since both target_uid and source_uid are used for the WHERE clause in the database queries, I'd suggest having an index for each of them and set uid alone as the primary key.

`comments_migrations.go` does this correctly and is fairly contained, to provide an example.

Reply:
> There can only be a single primary key, so what you're doing here is adding the primary key {uid, source_uid} which makes lookups that doesn't specify both (or at least not uid) inefficient.

having that composite primary key was intended as a way to enforce UID uniqueness by source UID and not across the whole table as we only care of their uniqueness given a source_uid. (and may help reducing UID collisions when generating them - even if unlikely to happen anyway).

I also thought about indices, which is something i wanted to better look at when adding the read endpoints.
given we want to get:
- all correlations
- all correlations originating from a given source_uid
- a single correlation (identified by source_uid+uid)
i was planning to add index in that PR for `source_uid` and `uid` (as a way to overcome the inefficiencies described by you).

wdyt?

Reply:
Ah, gotcha, that all sounds good to me!